From 61c43829dfeb41f8a849108b23688d25cd9014c8 Mon Sep 17 00:00:00 2001
From: Borislav Petkov <bp@suse.de>
Date: Mon, 19 May 2014 20:38:03 +0200
Subject: [PATCH] MCE, AMD: Check for userspace agent before decoding
Patch-mainline: never
References: bnc#871881

Now that mcelog supports AMD, we want to check whether it is running in
userspace, and, if so, to not decode to dmesg. The moment mcelog gets
unloaded or killed or b0rked, we fallback to decoding to dmesg.

Signed-off-by: Borislav Petkov <bp@suse.de>
---
 arch/x86/include/asm/mce.h       | 1 +
 arch/x86/kernel/cpu/mcheck/mce.c | 6 ++++++
 drivers/edac/mce_amd.c           | 3 +++
 3 files changed, 10 insertions(+)

Index: linux-3.0-SLE11-SP4/arch/x86/include/asm/mce.h
===================================================================
--- linux-3.0-SLE11-SP4.orig/arch/x86/include/asm/mce.h	2015-01-27 11:10:10.334361987 +0100
+++ linux-3.0-SLE11-SP4/arch/x86/include/asm/mce.h	2015-01-27 11:10:10.422361988 +0100
@@ -255,5 +255,7 @@
 extern void apei_mce_report_mem_error(int corrected,
 				      struct cper_sec_mem_err *mem_err);
 
+int mce_get_userspace_consumers_count(void);
+
 #endif /* __KERNEL__ */
 #endif /* _ASM_X86_MCE_H */
Index: linux-3.0-SLE11-SP4/arch/x86/kernel/cpu/mcheck/mce.c
===================================================================
--- linux-3.0-SLE11-SP4.orig/arch/x86/kernel/cpu/mcheck/mce.c	2015-01-27 11:10:10.334361987 +0100
+++ linux-3.0-SLE11-SP4/arch/x86/kernel/cpu/mcheck/mce.c	2015-01-27 11:13:04.114364750 +0100
@@ -1731,6 +1731,12 @@
 	return 0;
 }
 
+int mce_get_userspace_consumers_count(void)
+{
+	return open_count;
+}
+EXPORT_SYMBOL_GPL(mce_get_userspace_consumers_count);
+
 static void collect_tscs(void *data)
 {
 	unsigned long *cpu_tsc = (unsigned long *)data;
Index: linux-3.0-SLE11-SP4/drivers/edac/mce_amd.c
===================================================================
--- linux-3.0-SLE11-SP4.orig/drivers/edac/mce_amd.c	2015-01-27 11:10:10.418361988 +0100
+++ linux-3.0-SLE11-SP4/drivers/edac/mce_amd.c	2015-01-27 11:10:10.426361988 +0100
@@ -737,6 +737,9 @@
 	struct cpuinfo_x86 *c = &cpu_data(m->extcpu);
 	int ecc;
 
+	if (mce_get_userspace_consumers_count())
+		return NOTIFY_STOP;
+
 	if (amd_filter_mce(m))
 		return NOTIFY_STOP;
 
